src/stream-node.js

```diff
@@ -0,0 +1,60 @@
+import { PassThrough } from 'node:stream';
+import { renderToChunks } from './lib/chunked.js';
+
+/**
+ * @typedef {object} RenderToPipeableStreamOptions
+ * @property {() => void} [onShellReady]
+ * @property {() => void} [onAllReady]
+ * @property {(error) => void} [onError]
+ */
+
+/**
+ * @param {import('preact').VNode} vnode
+ * @param {RenderToPipeableStreamOptions} options
+ * @param {any} [context]
+ * @returns {{}}
+ */
+export function renderToPipeableStream(vnode, options, context) {
+	const encoder = new TextEncoder('utf-8');
+
+	const controller = new AbortController();
+	const stream = new PassThrough();
+
+	renderToChunks(vnode, {
+		context,
+		abortSignal: controller.signal,
+		onError: (error) => {
+			if (options.onError) {
+				options.onError(error);
+			}
+			controller.abort(error);
+		},
+		onWrite(s) {
+			stream.write(encoder.encode(s));
+		}
+	})
+		.then(() => {
+			options.onAllReady && options.onAllReady();
+			stream.end();
+		})
+		.catch((error) => {
+			stream.destroy(error);
+		});
+
+	Promise.resolve().then(() => {
+		options.onShellReady && options.onShellReady();
+	});
```

Comment:
Would this fire when the rendering of the shell already errors? Not sure if the expected behavior is to call `onShellReady` regardless or only in the error case.

Reply:
The render runs synchronously, so after a micro-tick this just tells us that the synchronous shell render is done